workload: introduce timeout for pre-warming connection pool

sean- opened the pull request:
Interrupting target instances during prewarming shouldn't cause workload to proceed: introduce a timeout to prewarming connections.  Connections will have 15s to 5min to warmup before the context will expire.

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101786)
<!-- Reviewable:end -->

tbg commented on pkg/workload/pgx_helpers.go:
```diff
@@ -319,32 +315,57 @@ func (m *MultiConnPool) WarmupConns(ctx context.Context, numConns int) error {
-				warmupConnsPerPool[i] = poolMaxConns
+			warmupPools[i].pool = p
+			warmupPools[i].maxConns = numPoolConns[i]
+			maxPoolConns := int(p.Config().MaxConns)
```
This isn't new code, but I'm wondering why we are not passing `MaxConns` into `distribute` but instead are allowing `distribute` to generate numbers that are larger than what we want, which will then result in the total number of conns being lower than `numConns`?

tbg commented on pkg/workload/pgx_helpers.go:
```diff
@@ -319,32 +315,57 @@ func (m *MultiConnPool) WarmupConns(ctx context.Context, numConns int) error {
+	for i, p := range warmupPools {
+		numWarmupConns += p.maxConns
+		if p.maxConns <= 0 {
+			warmupPools[i].maxConns = 1
```
Can we fix this up in the previous for loop? As is, I'm confused by why we add `p.maxConns` to `numWarmupConns` *before* then overwriting `p.maxConns` with 1. I also don't understand why `distribute` would ever return a zero for a pool. Can we not fold all of these semantics into `distribute` and simplify the code here?

tbg commented on pkg/workload/pgx_helpers.go:
```diff
@@ -319,32 +315,57 @@ func (m *MultiConnPool) WarmupConns(ctx context.Context, numConns int) error {
+	switch {
+	case int(warmupTime) < numWarmupConns*int(maxTimePerConn):
+		warmupTime = time.Duration(numWarmupConns * int(maxTimePerConn))
+	case warmupTime > maxWarmupTime:
```
Isn't this just `minWarmupTime > maxWarmupTime` which is `15s > 5m` and so this just evaluates to false? Don't you want to do this adjustment after having adjusted `warumTime`? If we enter the `case` on line 355 we _won't_ enter the one on 357[^1].

[^1]: https://play.golang.com/p/3w3m6qyGedh

tbg commented on pkg/workload/pgx_helpers.go:
```diff
@@ -319,32 +315,57 @@ func (m *MultiConnPool) WarmupConns(ctx context.Context, numConns int) error {
+	case warmupTime > maxWarmupTime:
+		warmupTime = maxWarmupTime
+	}
+	warmupCtx, cancel := context.WithTimeout(ctx, warmupTime)
```
nit: preference for just `ctx, cancel := ...` because that's the only context you'll be using, no need to keep the old one in scope.

tbg commented on pkg/workload/pgx_helpers.go:
```diff
@@ -357,8 +378,6 @@ func (m *MultiConnPool) WarmupConns(ctx context.Context, numConns int) error {
-			// NOTE(seanc@): Release() connections before canceling the warmupCtx to
-			// prevent partially established connections from being Acquire()'ed.
 			conn.Release()
 		}
```
I can't comment on it but line 395 below (`break WARMUP`) is dead code. When a `Done()` channel is closed, `Err()` is guaranteed to return an error (that's in the contract) so you just need a blanket `return err` inside of the `case`.

tbg reviewed (changes requested):
Still have some confusions and I think it could be simplified a bit, in particular around the `distribute()` function.

Also, meta question, don't we want pre-warming to be optional? `./cockroach workload run X ` has the flag `--tolerate-errors`. Ideally we want to be able to run it against a cluster with a node down (but some node reachable) and it would start up. We shouldn't insist on all provided connection strings working.
I think the old behavior was the same, though. I think `--tolerate-errors` only ever worked reliable if the errors occurred "late enough", when the workload was properly running. Still, food for thought.

renatolabs commented:
Friendly ping.

We have a test that frequently times out in connection warmup (after 1h30m) with `workload run tpcc --tolerate-errors` (#102687).

sean- reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @tbg)

___
*[`pkg/workload/pgx_helpers.go` line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101786#gh-1178727757:-NUxBebKD0Tws2Fu98wC:brroh78) ([raw file](https://github.com/cockroachdb/cockroach/blob/9373f7c5543d854f81005a329ae22cc08d627544/pkg/workload/pgx_helpers.go#L333)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This isn't new code, but I'm wondering why we are not passing `MaxConns` into `distribute` but instead are allowing `distribute` to generate numbers that are larger than what we want, which will then result in the total number of conns being lower than `numConns`?
</blockquote></details>

We automatically set the connection count if a user specifies `0` connections.  In this case, the user specified a specific number of connections, which we will then distribute them across the supplied URLs.  We still set a per-pool max connection count.  If a user specifies a max connection count of 100, only passes in a handful of URLs, and then sets the number of connections to 10K, we will clamp the number of connections to 100 because that was the max that a user specified.  Arguably, there's a bug here where this should only be applied if `maxPoolConns` is also greater than zero.
___
*[`pkg/workload/pgx_helpers.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101786#gh-1178729906:-NUxfyZ79HQ-G5JMdTxv:bs8oxz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9373f7c5543d854f81005a329ae22cc08d627544/pkg/workload/pgx_helpers.go#L344)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can we fix this up in the previous for loop? As is, I'm confused by why we add `p.maxConns` to `numWarmupConns` *before* then overwriting `p.maxConns` with 1. I also don't understand why `distribute` would ever return a zero for a pool. Can we not fold all of these semantics into `distribute` and simplify the code here?
</blockquote></details>

Please take a look at the updated revision.  I didn't like how this was earlier, either, and think this is easier to follow.
___
*[`pkg/workload/pgx_helpers.go` line 357 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101786#gh-1178733185:-NV0hRso2q3oG_Rns1fX:b-p3yyd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9373f7c5543d854f81005a329ae22cc08d627544/pkg/workload/pgx_helpers.go#L357)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Isn't this just `minWarmupTime > maxWarmupTime` which is `15s > 5m` and so this just evaluates to false? Don't you want to do this adjustment after having adjusted `warumTime`? If we enter the `case` on line 355 we _won't_ enter the one on 357[^1].

[^1]: https://play.golang.com/p/3w3m6qyGedh
</blockquote></details>

It is.  This `switch` block used to have 3x `case` in it, and I prefer using `switch`+`case` over `if else` blocks.  The gist of this code block is to create a sliding scale for the warmup time with a min and max, though this wasn't clamping things.  Updated.


<!-- Sent from Reviewable.io -->

sean- committed:
workload: introduce timeout for pre-warming connection pool

Interrupting target instances during prewarming shouldn't cause workload to
proceed: introduce a timeout to prewarming connections.  Connections will have
15s to 5min to warmup before the context will expire.

Epic: none
Release note: None

sean- reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @tbg)

___
*[`pkg/workload/pgx_helpers.go` line 382 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101786#gh-1178735688:-NV11NIX34vTdnBULmYC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9373f7c5543d854f81005a329ae22cc08d627544/pkg/workload/pgx_helpers.go#L382)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I can't comment on it but line 395 below (`break WARMUP`) is dead code. When a `Done()` channel is closed, `Err()` is guaranteed to return an error (that's in the contract) so you just need a blanket `return err` inside of the `case`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sean- committed:
workload: log a warning instead of erroring when warming up connections

Epic: none
Release note: None
Fixes: #102687

sean- reviewed (commented):
@tbg I agree pre-warming should be optional and was trying not to tinker with the original thrust of the code.

@renatolabs : this will provide an upper bound for establishing connections, which will be helpful in flakey environments.

I'm not opposed to making connection prewarming a soft error, in that a failure in warming up connections will allow `workload` to continue.  The expectation that 100% of all connections will succeed when benchmarking a distributed system is unrealistic when running workers with instance preemption.  I've pushed a small change that changes `WarmupConns()` to log on error instead of exit.  Plumbing `-tolerate-errors` down is doable but seems heavy-weight.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786#-:-NV11rKJ4a5A07jSjwnS:b-qcir2z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @tbg)



<!-- Sent from Reviewable.io -->

sean- reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @tbg)

___
*[`pkg/workload/pgx_helpers.go` line 217 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101786#-NV192XxAJmwNdU22m8u:-NV192Xy2h5DbVte_K78:b-4p6tdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f7143312ceff9b17ac77053a643677160a4fd3f/pkg/workload/pgx_helpers.go#L217)):*
> ```Go
> 
> 	if err := m.WarmupConns(ctx, cfg.WarmupConns); err != nil {
> 		log.Warningf(ctx, "warming up connection pool failed (%v), continuing workload", err)
> ```

@tbg / @renatolabs : this is a change in behavior, but I think would help our overall testing with the new max connection timeouts added elsewhere in this PR.

CC @srosenberg 


<!-- Sent from Reviewable.io -->

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. I'll make sure this fixes the failures I've been seeing on roachtests (it likely will, especially now that we don't return an error if warmup times out). @tbg should also take another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786#-:-NV1T7VI6ya0uKYeGIBV:b-skzpra)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @sean-, @srosenberg, and @tbg)

___
*[`pkg/workload/pgx_helpers.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101786#gh-1178729906:-NV1RAw2E1l_biwSVaak:b-3qyplj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9373f7c5543d854f81005a329ae22cc08d627544/pkg/workload/pgx_helpers.go#L344)):*
<details><summary><i>Previously, sean- (Sean Chittenden) wrote…</i></summary><blockquote>

Please take a look at the updated revision.  I didn't like how this was earlier, either, and think this is easier to follow.
</blockquote></details>

This code also suggests `poolMaxConns` is never `<= 0`. Otherwise, we could be setting `maxConns` to `0` here, which is probably not valid.
___
*[`pkg/workload/pgx_helpers.go` line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101786#-NV1Q-SoDq8Qz9VHhWLG:-NV1Q-SoDq8Qz9VHhWLH:bqndz3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/de709fab1eb5409883afa8ea3f8e411afb161ece/pkg/workload/pgx_helpers.go#L292)):*
> ```Go
> 		// Tune max conns for the pool
> 		switch {
> 		case totalNumConns == 0 && poolMaxConns > 0:
> ```

Can `poolMaxConns` be `<= 0`? According to the documentation, the default is `the greater of 4 or runtime.NumCPU()`, and custom values are `integer greater than 0`.

If `pgxpool` guarantees that invariant, we could get rid of the `case` below and simplify this logic (took me a while to understand).
___
*[`pkg/workload/pgx_helpers.go` line 217 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101786#-NV192XxAJmwNdU22m8u:-NV1S_KW2QsT56DXi6dH:b-a7ljwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f7143312ceff9b17ac77053a643677160a4fd3f/pkg/workload/pgx_helpers.go#L217)):*
<details><summary><i>Previously, sean- (Sean Chittenden) wrote…</i></summary><blockquote>

@tbg / @renatolabs : this is a change in behavior, but I think would help our overall testing with the new max connection timeouts added elsewhere in this PR.

CC @srosenberg 
</blockquote></details>

Makes sense to me.


<!-- Sent from Reviewable.io -->

sean- reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @tbg)

___
*[`pkg/workload/pgx_helpers.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101786#gh-1178729906:-NV1pO127kuaCqvfHKwr:bh766tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9373f7c5543d854f81005a329ae22cc08d627544/pkg/workload/pgx_helpers.go#L344)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This code also suggests `poolMaxConns` is never `<= 0`. Otherwise, we could be setting `maxConns` to `0` here, which is probably not valid.
</blockquote></details>

If we set `maxConns` to `0` here, we rely on the pgxpool defaults.
___
*[`pkg/workload/pgx_helpers.go` line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101786#-NV1Q-SoDq8Qz9VHhWLG:-NV1oVCf7M5nFqmev7p9:b-b7qu6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/de709fab1eb5409883afa8ea3f8e411afb161ece/pkg/workload/pgx_helpers.go#L292)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can `poolMaxConns` be `<= 0`? According to the documentation, the default is `the greater of 4 or runtime.NumCPU()`, and custom values are `integer greater than 0`.

If `pgxpool` guarantees that invariant, we could get rid of the `case` below and simplify this logic (took me a while to understand).
</blockquote></details>

`poolMaxConns` is passed down from the CLI, so it's possible, yes.  pgxpool does not guarantee these minimums.  The default, however, uses this heuristic: https://github.com/jackc/pgx/blob/master/pgxpool/pool.go#L289-L304


<!-- Sent from Reviewable.io -->

renatolabs commented:
Confirming the changes here fix the failures I've been seeing in roachtests.

tbg reviewed (approved):
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101786)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan and @renatolabs)



<!-- Sent from Reviewable.io -->

sean- commented:
bors r+

sean- commented:
bors r+
